src/java/org/apache/cassandra/config/DatabaseDescriptor.java

```diff
@@ -1499,20 +1500,24 @@ private static DiskAccessMode resolveCommitLogWriteDiskAccessMode(DiskAccessMode
             providedDiskAccessMode = compressOrEncrypt ? DiskAccessMode.standard : DiskAccessMode.mmap;
         }
 
-        return providedDiskAccessMode;
+        return Pair.create(providedDiskAccessMode, directIOSupported);
     }
 
-    private static void validateCommitLogWriteDiskAccessMode(DiskAccessMode diskAccessMode) throws ConfigurationException
+    private static void validateCommitLogWriteDiskAccessMode(Pair<DiskAccessMode, Boolean> accessModeDirectIoPair) throws ConfigurationException
     {
         boolean compressOrEncrypt = getCommitLogCompression() != null || (getEncryptionContext() != null && getEncryptionContext().isEnabled());
 
-        if (compressOrEncrypt && diskAccessMode != DiskAccessMode.standard)
+        if (!accessModeDirectIoPair.right && accessModeDirectIoPair.left == DiskAccessMode.direct)
         {
-            throw new ConfigurationException("commitlog_disk_access_mode = " + diskAccessMode + " is not supported with compression or encryption. Please use 'auto' when unsure.", false);
+            throw new ConfigurationException("commitlog_disk_access_mode can not be set to direct when direct IO is not supported now.");
```

Comment:
Nits: 

I would remove "now" from the message, or replace it with "by the file system".

I personally would just throw the exception inside `resolveCommitLogWriteDiskAccessMode` to avoid returning pairs, but I'm happy to accept this version as well.